### PR TITLE
Fix OOME: Metaspace

### PR DESCRIPTION
**Pull request by vmishenev**
#1405 
100 tasks, MaxMetaspaceSize=300m
image
image

### CODE REVIEW

File: integration-tests/gradle/src/integrationTest/kotlin/org/jetbrains/dokka/it/gradle/Stress0IntegrationTest.kt [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package org.jetbrains.dokka.it.gradle
+
+import org.gradle.testkit.runner.TaskOutcome
+import org.junit.runners.Parameterized.Parameters
+import java.io.File
+import kotlin.test.*
+
+class Stress0IntegrationTest(override val versions: BuildVersions) : AbstractGradleIntegrationTest() {
```
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply integration-tests/gradle/src/integrationTest/kotlin/org/jetbrains/dokka/it/gradle/Stress0IntegrationTest.kt:8]: Also, can you please add a KDoc for this test, explaining what it does and under which conditions it fails?

Something along the lines of 

> Creates 100 tasks for the test project and runs them sequentially under low memory settings. 
> 
> If the test passes, it's likely there are no noticeable memory leaks. If it fails, it's likely that memory is leaking somewhere.

**Review by IgnatBeresnev — COMMENTED** [review]
Sorry for so many nitpicks :sweat_smile:

File: integration-tests/gradle/src/integrationTest/kotlin/org/jetbrains/dokka/it/gradle/Stress0IntegrationTest.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        templateProjectDir.listFiles().orEmpty()
+            .filter { it.isFile }
+            .forEach { topLevelFile -> topLevelFile.copyTo(File(projectDir, topLevelFile.name)) }
+
+        File(templateProjectDir, "src").copyRecursively(File(projectDir, "src"))
+    }
+
+    @Test
+    fun execute() {
+       //if(!TestEnvironment.isExhaustive) return
```
[the remark this answers is not in the document]
Comment (vmishenev) [reply integration-tests/gradle/src/integrationTest/kotlin/org/jetbrains/dokka/it/gradle/Stress0IntegrationTest.kt]: I removed the condition, but it makes sense to add it after the support of `isExhaustive` flag.

File: integration-tests/gradle/projects/it-sequential-tasks-execution-stress/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    id("org.jetbrains.dokka")
+}
+
+buildscript {
+    dependencies {
+        classpath("org.jetbrains.dokka:dokka-base:${System.getenv("DOKKA_VERSION")}")
+    }
+}
+
+version = "1.7.10-SNAPSHOT"
```
Comment (IgnatBeresnev) [inline integration-tests/gradle/projects/it-sequential-tasks-execution-stress/build.gradle.kts:19]: 1.7.20?

File: runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/AbstractDokkaTask.kt [hunk 4]
```diff
@@ -89,7 +89,14 @@ abstract class AbstractDokkaTask : DefaultTask() {
     internal open fun generateDocumentation() {
         DokkaBootstrap(runtime, DokkaBootstrapImpl::class).apply {
             configure(buildDokkaConfiguration().toJsonString(), createProxyLogger())
-            generate()
+            /**
+             * Run in a new thread to avoid memory leaks that are related to ThreadLocal (that keeps `URLCLassLoader`)
+             * Currently, all `ThreadLocal`s are in a compiler/IDE codebase.
```
Comment (IgnatBeresnev) [inline runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/AbstractDokkaTask.kt:94]:
```suggestion
             * Currently, all leaking `ThreadLocal`s are in the compiler/IDE codebase.
```

**Review by IgnatBeresnev — APPROVED** [review]
:+1:

**Pull request merged**